proto/sql_api.proto

```diff
@@ -0,0 +1,154 @@
+// clients can submit the same command multiple times and always
+// receive the same response. This is common on retries over flaky
+// networks. However, the system imposes a limit on how long
+// idempotence is provided. Retries over an hour old are not
+// guaranteed idempotence and may be executed more than once with
+// potentially different results.
+//
+// ClientCmdID contains the client's timestamp and a client-generated
+// random number. The client Timestamp is specified in unix
+// nanoseconds and is used for some uniqueness but also to provide a
+// rough ordering of requests, useful for data locality on the
+// server. The Random is specified for additional uniqueness.
+// NOTE: An accurate time signal IS NOT required for correctness.
+message ClientCmdID {
+  // Nanoseconds since Unix epoch.
+  optional int64 wall_time = 1 [(gogoproto.nullable) = false];
+  optional int64 random = 2 [(gogoproto.nullable) = false];
+}
+
+// ReadConsistencyType specifies what type of consistency is observed
+// during read operations.
+enum ReadConsistencyType {
+  option (gogoproto.goproto_enum_prefix) = false;
+  // CONSISTENT reads are guaranteed to read committed data; the
+  // mechanism relies on clocks to determine lease expirations.
+  CONSISTENT = 0;
+  // CONSENSUS requires that reads must achieve consensus. This is a
+  // stronger guarantee of consistency than CONSISTENT.
+  //
+  // TODO(spencer): current unimplemented.
+  CONSENSUS = 1;
+  // INCONSISTENT reads return the latest available, committed values.
+  // They are more efficient, but may read stale values as pending
+  // intents are ignored.
+  INCONSISTENT = 2;
+}
+
+// CmdRequestHeader is supplied with every CmdRequest.
+message CmdRequestHeader {
+  // Timestamp specifies time at which read or writes should be
+  // performed. If the timestamp is set to zero value, its value
+  // is initialized to the wall time of the receiving gateway node.
+  optional Timestamp timestamp = 3 [(gogoproto.nullable) = false];
+  // CmdID is optionally specified for request idempotence
+  // (i.e. replay protection).
+  optional ClientCmdID cmd_id = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "CmdID"];
+  // User is the originating user. Used to lookup priority when
+  // scheduling queued operations at target node.
+  optional string user = 5 [(gogoproto.nullable) = false];
+  // UserPriority specifies priority multiple for non-transactional
+  // commands. This value should be a positive integer [1, 2^31-1).
+  // It's properly viewed as a multiple for how likely this
+  // transaction will be to prevail if a write conflict occurs.
+  // Commands with UserPriority=100 will be 100x less likely to be
+  // aborted as conflicting transactions or non-transactional commands
+  // with UserPriority=1. This value is ignored if Txn is
+  // specified. If neither this value nor Txn is specified, the value
+  // defaults to 1.
+  optional int32 user_priority = 6 [default = 1];
```

Comment:
`client.DB.Open` currently allows setting the priority in the URL: https://github.com/cockroachdb/cockroach/blob/master/client/db.go#L180. I was thinking drivers could do something similar.

@bdarnell Any thoughts on how we implement connection specific values when we don't actually have a persistent connection? Something that has been banging around my head is that a `SET` operation would cause an opaque blob with the "connection values" to be returned in the response which the client in turn would send on subsequent requests.